js/localization/ldml/date.parser.js

```diff
@@ -40,7 +41,7 @@ const PATTERN_REGEXPS = {
         return dateParts.getPeriodNames(FORMAT_TYPES[count < 3 ? 3 : count], 'format').join('|');
     },
     d: function(count) {
-        return '0?[1-9]|[12][0-9]|3[01]';
+        return count === 2 ? '3[01]|[12][0-9]|0[1-9]' : '0?[1-9]|[12][0-9]|3[01]';
```

Comment:
There is only one sign difference, but it is not obious because of different parts oder. I recommend to apply the condition for only one that sign.

Reply:
Why did you make more strict regexp only for two of date parts: month and day. What about the rest of them?